base/XGamepad.ts

```diff
@@ -0,0 +1,503 @@
+            case Button.LEFT:
+                return Button.RIGHT;
+            case Button.RIGHT:
+                return Button.LEFT;
+            case Button.UP:
+                return Button.DOWN;
+            case Button.DOWN:
+                return Button.UP;
+            case Button.L:
+                return Button.R;
+            case Button.R:
+                return Button.L;
+            default:
+                return null;
+        }
+    }
+
+    /** Every game buttons of the emulated gamepad */
+    buttons: {[button in Button | CButton]?: GamepadButton} = [];
+    /** The stick dead zone */
+    stick_dead_zone: number;
+    /** The trigger dead zone */
+    trigger_dead_zone: number;
+    /** Whether the last button pressed comes from the keybord or the controller */
+    is_last_input_gamepad: boolean;
+
+    /**
+     * @param {GoldenSun} data - Master game data object
+     */
+    constructor(data: GoldenSun) {
+        Gamepad.initialize(data, navigator.language === "fr-FR" ? "azerty_inputs" : "default_inputs");
+        this.register_handle_events(data.game);
+        Buttons.forEach(button_name => {
+            this.buttons[Button[button_name]] = new GamepadButton(this, Button[button_name]);
+        });
+        CButtons.filter(bn => bn).forEach(button_name => {
+            this.buttons[CButton[button_name]] = new GamepadButton(this, CButton[button_name]);
+        });
+        const mirror_button = (button: Button | CButton, to: Button | CButton) => {
+            this.buttons[button].on_down.add(() => this.on_gamepad_down(to));
+            this.buttons[button].on_up.add(() => this.on_gamepad_up(to));
+        };
+        if (data.dbs.init_db.gamepad?.use_trigger_as_button === true) {
+            mirror_button(CButton.LT, Button.L);
+            mirror_button(CButton.RT, Button.R);
+        }
+        if (data.dbs.init_db.gamepad?.left_stick_as_dpad === true) {
+            mirror_button(CButton.LLEFT, Button.LEFT);
+            mirror_button(CButton.LRIGHT, Button.RIGHT);
+            mirror_button(CButton.LUP, Button.UP);
+            mirror_button(CButton.LDOWN, Button.DOWN);
+        }
+        if (data.dbs.init_db.gamepad?.right_stick_as_dpad === true) {
+            mirror_button(CButton.RLEFT, Button.LEFT);
+            mirror_button(CButton.RRIGHT, Button.RIGHT);
+            mirror_button(CButton.RUP, Button.UP);
+            mirror_button(CButton.RDOWN, Button.DOWN);
+        }
+        this.stick_dead_zone = data.dbs.init_db.gamepad?.stick_dead_zone ?? 0.5;
+        this.trigger_dead_zone = data.dbs.init_db.gamepad?.trigger_dead_zone ?? 0.6;
+        this.is_last_input_gamepad = false;
+    }
+
+    /**
+     * Press the game button if it is up then triggers the listeners.
+     * @param {Button|CButton} game_button - The game button to press
+     * @param {?KeyboardEvent} event - The keyboard event if any
+     */
+    _on_down(game_button: Button | CButton, event?: KeyboardEvent) {
```

Comment:
This method seems to be `private`, if so, please use the `private` access modifier.